mirrord/protocol/src/codec.rs

```diff
@@ -142,77 +143,37 @@ pub enum DaemonMessage {
     SwitchProtocolVersionResponse(#[bincode(with_serde)] semver::Version),
 }
 
-pub struct ClientCodec {
+pub struct ProtocolCodec<I, O> {
     config: bincode::config::Configuration,
+    /// Phantom just to associate the message types with the struct.
+    /// Implementing a trait to associate the types is not good enough, because we can't blanket
+    /// implement the foreign `Encoder`/`Decoder` traits, so we have to implement them for a
+    /// struct.
```

Comment:
I'm being dumb, what do you mean here?

```rust
trait Codec {
  type In: Decode;
  type Out: Encode;
}
```

This wouldn't work? (not that there is anything wrong with the phantom approach btw, just want to get a better understanding)

Reply:
What I meant to say was that we can't implement `Codec` for  a generic `ClientCodec` and `DaemonCodec` structs to associate the correct message types with each of them and then just implement `Encoder` and `Decoder` for a generic `C: Codec`, because that `C` could be a foreign type and then we would be implementing a foreign trait (`Encoder`/`Decoder`) for a foreign type (anything that implements `Codec`). I think. But I now see that this comment is really not helpful at all and the code is clearer without it.
```suggestion
    /// Phantom fields to make this struct generic over message types.
```

Reply:
Hmm I think what confused me most is that, this change is geared towards turning this more generic to help with the protocol versioning thingy, right?

If that's the case, could you leave a comment saying that?

Reply:
Should we really comment on a change in code? Future readers will only see the new state, not the old one, so it seems weird to me to discuss that change in the code.
I having a generic `ProtocolVersion` struct is already justified by the fact that we don't implement the `Encoder` and `Decoder` traits twice each with the exact same code. Should I comment that above the struct?

Reply:
Nah, you're right. I was looking at this as a sort of "TODO: changing this for the protocol", but the change is already being made.